File: Exámenes/Septiembre2013/Resuelto/Haskell/DataStructures/Graph/Graph.hs
-------------------------------------------------------------------------------
-- Graph defined by list of vertices and adjancency function
--
-- Data Structures. Grado en Informática. UMA.
-- Pepe Gallardo, 2012
-------------------------------------------------------------------------------

module DataStructures.Graph.Graph   
  ( Graph
  , Edge
  , Path
  , mkGraphEdges   
  , mkGraphSuc   
  , successors
  , edges     
  , vertices
  , degree
  ) where

import Data.List(nub)

type Edge a = (a,a)

type Path a = [a] -- Path represented as list of vertices

data Graph a = G [a] (a -> [a])

mkGraphSuc :: [a] -> (a -> [a]) -> Graph a
mkGraphSuc vs sucs = G vs sucs

mkGraphEdges :: (Eq a) => [a] -> [Edge a] -> Graph a
mkGraphEdges vs es = G vs sucs
 where
   sucs v = nub $ [ y | (x,y) <- es, x==v] 
                 ++ 
                  [ x | (x,y) <- es, y==v]                

successors :: Graph a -> a -> [a]
successors (G vs sucs) v = sucs v 

vertices :: Graph a -> [a]
vertices (G vs sucs) = vs

edges :: (Eq a) => Graph a -> [Edge a]
edges (G vs sucs) = [ (v,w) | v <- vs, w <- sucs v ] 

degree :: Graph a -> a -> Int
degree g v = length (successors g v)

instance Show (Graph a) where
  show g = "Graph"

File: Exámenes/2016/Febrero/Java/eulerianCycle/src/dataStructures/graph/Graph.java
/**
 * @author Pepe Gallardo, Data Structures, Grado en Informática. UMA.
 *
 * Common interface for all graphs
 */

package dataStructures.graph;

import dataStructures.set.Set;

/**
 * Interface for undirected graphs.
 *
 * @param <V> Type of vertices in graph.
 */
public interface Graph<V> extends Traversable<V>, Cloneable {
	/**
	 * true if graph has no vertices.
	 */
	boolean isEmpty();

	/**
	 * Inserts a vertex in the graph. Vertices must be inserted
	 * into graph prior to adding incident edges.
	 *  
	 * @param v Vertex to add to graph.
	 */
	void addVertex(V v);
	
	/**
	 * Removes one vertex from the graph. It also removes all edges
	 * that are incident to removed vertex.
	 * @param v Vertex to remove.
	 */
	void deleteVertex(V v);

	
	/**
	 * Adds an edge to the graph. Note that vertices should be in
	 * 
	 * @param v
	 * @param w
	 */
	void addEdge(V v, V w);

	void deleteEdge(V v, V w);
	
	int numVertices();

	int numEdges();
	
	Set<V> vertices();

	int degree(V v);	
	
	Object clone();
	
}
